File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/Tree/235_LowestCommonAncestorOfBinarySearchTree.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# @Author: xuezaigds@gmail.com


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

c.. Solution o..
    # Easy to understand
    ___ lowestCommonAncestor  root, p, q
        min_val = min(p.val, q.val)
        max_val = m..(p.val, q.val)
        _____ root:
            value = root.val
            __ min_val <= value <= max_val:
                r_ root
            ____ max_val < value:
                root = root.left
            ____
                root = root.right
        r_ None


c.. Solution_2 o..
    """
    One elegant code, some puzzling but short code. according to:
    https://leetcode.com/discuss/44959/3-lines-with-o-1-space-1-liners-alternatives
    Just walk down from the whole tree's root as long as
    both p and q are in the same subtree
    """
    ___ lowestCommonAncestor  root, p, q
        _____ (root.val - p.val) * (root.val - q.val) > 0:
            root = (root.left, root.right)[p.val > root.val]
        r_ root

File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/Tree/235_LowestCommonAncestorOfBinarySearchTree.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-09-20 19:37:23
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
c.. Solution {
public:
    // Easy to understand
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        int min = p->val < q->val ? p->val : q->val;
        int m.. = p->val < q->val ? q->val : p->val;
        _____(root){
            __(root->val < min){
                root = root->right;
            }
            else __(root->val > m..){
                root = root->left;
            }
            else{
                r_ root;
            }
        }
        r_ nullptr;
    }
};
